drgn_tools/multipath.py

```diff
@@ -0,0 +1,345 @@
+    Format the output header as per ps policy.
+
+    :param  ps_type: Path selection policy name.
+
+    :return None.
+    """
+    header_mapping = {
+        "service-time": "   InFlightIO_Size   RThroughput  RCount  Load",
+        "round-robin": "   Repeat_Count",
+        "queue-length": "   Repeat_Count  Queue_Length",
+        "io-affinity": "   Refcount   failed",
+        "historical-service-time": "Feature not supported for "
+        "historical-service-time.",
+    }
+    header = header_mapping.get(ps_type)
+    if header is not None:
+        print(header)
+    else:
+        """
+        We shouldn't reach here.
+        """
+        print("\n\nPath selection policy %s not supported.\n\n" % ps_type)
+
+
+def switch_output(ps_type: str, pi: Object) -> None:
+    """
+    Format the path info output as per ps policy.
+
+    :param  ps_type: Path selection policy name.
+    :param pi: `struct path_info *`
+
+    :return None.
+    """
+    output_mapping = {
+        "service-time": "{:<6}{:<12d}{:<5}{:<7d}{:<3}{:<6d}{:<1}{:d}",
+        "round-robin": "{:<5}{:d}",
+        "queue-length": "{:<5}{:<9d}{:<5}{:d}",
+        "io-affinity": "{:<4}{:<7d}{:<3}{:d}",
+        "historical-service-time": "",
+    }
+    output_format = output_mapping.get(ps_type)
+    if output_format:
+        if ps_type == "service-time":
+            load = pi.in_flight_size.counter / pi.relative_throughput
+            print(
+                output_format.format(
+                    space,
+                    pi.in_flight_size.counter.value_(),
+                    space,
+                    pi.relative_throughput.value_(),
+                    space,
+                    pi.repeat_count.value_(),
+                    space,
+                    load.value_(),
+                )
+            )
+        elif ps_type == "round-robin":
+            print(output_format.format(space, pi.repeat_count.value_()))
+        elif ps_type == "queue-length":
+            print(
+                output_format.format(
+                    space,
+                    pi.repeat_count.value_(),
+                    space,
+                    pi.qlen.counter.value_(),
+                )
+            )
+        elif ps_type == "io-affinity":
+            print(
+                output_format.format(
+                    space,
+                    pi.refcount.refs.counter.value_(),
+                    space,
+                    pi.failed.value_(),
+                )
+            )
+        else:
+            print(output_format)
+    else:
+        print("\n\nPath selection policy %s not supported.\n\n" % ps_type)
+
+
+def disk_size_and_permission(bdev: Object) -> None:
+    """
+    Print Disk write permission.
+    Print Disk Size in bytes.
+
+    :param bdev: `struct block device *`
+
+    :returns: None
+    """
+    if blkdev_ro(bdev):
+        print("permission = r")
+    else:
+        print("permission = rw")
+    print("Size = %u" % bdev.bd_inode.i_size)
```

Comment:
Updated.

Reply:
I do not have UEK5 setup which has multipath devices. I am using a simulated environment. And it seems this is causing some pointer in block_devices to be null.  For read only case bd_part is null in block_device for uek4 and uek5. Which is causing blkdev_ro() to fail with error.

`--------------- Device [1] ---------------
mpathe (333333330000007d6) dm-7
struct multipath * (0xffff9b1069ffc600)
Flags = 0x0
Current path = 8:96
hw_handler_name = 0
struct mapped_device * 0xffff9b0edab00000
error: Encountered exception in multipath:
Traceback (most recent call last):
  File "/drgn_home/shmsingh/drgn-tools/drgn_tools/corelens.py", line 340, in _run_module
    mod.run(prog, args)
  File "/drgn_home/shmsingh/drgn-tools/drgn_tools/multipath.py", line 305, in run
    show_mp(prog)
  File "/drgn_home/shmsingh/drgn-tools/drgn_tools/multipath.py", line 293, in show_mp
    dump_dm(prog, dm, name, uuid)
  File "/drgn_home/shmsingh/drgn-tools/drgn_tools/multipath.py", line 273, in dump_dm
    show_multipath(prog, target, dm)
  File "/drgn_home/shmsingh/drgn-tools/drgn_tools/multipath.py", line 245, in show_multipath
    print("Permission = %s" % ("r" if blkdev_ro(bdev) else "rw"))
  File "/drgn_home/shmsingh/drgn-tools/drgn_tools/block.py", line 59, in blkdev_ro
    return bdev.bd_part.policy != 0
_drgn.FaultError: could not read memory from kdump: Cannot get page I/O address: PDPT table not present: p4d[0] = 0x0: 0x318
`

Reply:
UEK6 output snippet:
----------------------

--------------- Device [18] ---------------
mpathf (3600144f0d06e755600005ca3c5ae0002) dm-19
struct multipath * (0xffff985cb7602800)
Flags = MPATHF_QUEUE_IF_NO_PATH|MPATHF_SAVED_QUEUE_IF_NO_PATH
Current path = 8:112
hw_handler_name = alua
struct mapped_device * 0xffff985ea48f0000
Permission = rw
Size = 2147483648
Policy = service-time
Path_No.  Disk_Name  DevName     State        InFlightIO_Size   RThroughput  RCount  Load
   1        sdh       8:112      active             0                1         1      0

Reply:
I am not sure what "simulated environment" is, i will be good to have two env, one with ol7 and one with ol8, you can use ol7 env to test uek4/5 and ol8 env to test uek6/7. Then you can investigate what cause blkdev_ro fail.

Reply:
@biger410  by simulated environment I mean, I am using scsci_debug module to create scsi disks with multipath in RAM, which I am using as my test disks.

I will check with Rajan if we can install OL7 in one of the test server. We have 3 server that I know have multipath enabled disks, but all are with OL8/OL9. 

For the time being, while I am working on the test environment I have pushed my changes for code review.  

Gitlab tests have failed with the same error I posted above since gitlab also use the same vmcore for UEK4 and UEK5.